_posts/2019-09-06-kubernetes-and-hokusai.md

```diff
@@ -0,0 +1,146 @@
+thrown around a few times, but had no idea how it worked or what it was good for.
+
+Kubernetes is still a bit of a mystery to me, but I'm able to do a lot of Kubernetes operations quickly and easily
+thanks to an open-source tool developed at Artsy: [Hokusai](https://github.com/artsy/hokusai).
+
+In this post, I'll give some background on Kubernetes, a brief history of Hokusai and a description of its
+functionality, and some pointers for how to get started using it.
+
+<!-- more -->
+
+# What is Kubernetes?
+
+On a high level, Kubernetes is a tool designed to _orchestrate containers at scale._
+
+Let's break that down a bit. First, some helpful vocab:
+
+**Container**: Effectively code + all necessary dependencies for an application. A
+["standardized unit of software"](https://www.docker.com/resources/what-container).
+
+**Pods**: Single instance of an application or process. Consists of one or more containers, though one container
+per pod is the most common use case.
+
+**Replicas**: Multiple instances of the same application, each running on its own pod.
+
+**Node**: A physical or virtual machine that runs a pod or pods.
+
+**Cluster**: A node or nodes managed by a "master" machine (which could be a physical computer or a virtual
+machine - it's one node within the cluster).
+
+**Container orchestration**: A systemized approach to managing containers. Allows for things like auto-scaling,
+easy rollouts and rollbacks, and automation of container downtime (i.e. something goes wrong with a machine and
+causes your app to crash; a new container gets spun up immediately so that your app doesn't go down).
+
+Sources: [Kubernetes docs](https://kubernetes.io/docs/concepts/overview/what-is-kubernetes/),
+[Infoworld](https://www.infoworld.com/article/3268073/what-is-kubernetes-your-next-application-platform.html),
+[Docker](https://www.docker.com/resources/what-container)
+
+To sum up the general structure: a Kubernetes cluster contains nodes which contain pods which contain containers.
+Not at all confusing, right?
+
+Kubernetes, in a general sense, allows you to configure the containers in which your application will run. With a
+properly configured Kubernetes cluster, this makes it easy to scale applications up or down as needed to deal with
+traffic patters, maintain a zero-downtime deployment, and more. Very cool.
+
+# What is Hokusai?
+
+When Artsy's Engineering team was contemplating a move to Kubernetes from Heroku, we had beef with a few things.
+
+For one, we wanted to be able to do a few core things simply and easily using the command line. While Kubernetes
+has robust CLI tooling using [kubectl](https://kubernetes.io/docs/reference/kubectl/overview/), it's also very
+complex. We wanted to be able to quickly and easily do the things we were used to doing with Heroku; we preferred
+`heroku logs` to `kubectl logs [POD]` (where we would have to either look up or know the specific pod name we
```

Comment:
👏 you nailed it!  That was a major design decision - keeping configuration and furthermore an application developer's interactions with their application in the app's repo